docs: add docs on machine join config and kernel parameters (#143)
siderolabs/docs: public/omni/omni-cluster-setup/registering-machines/join-machines-to-omni.mdx

---
title: Join machines to Omni
---

> Omni lets you connect existing Talos nodes—whether they’re running on bare metal, in QEMU, or on cloud instances—to your Omni environment so it can automatically take over their management.

Reviewer: ```suggestion
Omni lets you connect existing Talos nodes—whether they’re running on bare metal or as VMs—to your Omni environment so it can automatically take over their management.
```

> You can register a node by applying a **Machine Join Config**, which tells the node how to connect to Omni during startup.

Reviewer: ```suggestion
You can register a node to Omni in 2 different ways.

1. Boot a machine with pre-configured SideroLink configuration from Omni or the Image Factory.
2. Boot a standard Talos Linux installation media and provide a Machine Join Config
```

> **Note:** Once a machine joins Omni, it is managed entirely through Omni.  

> > The local Talos API is disabled for security reasons, and all future configuration changes must be made through Omni.

Reviewer: ```suggestion
> The local Talos API is disabled for security reasons, and all future configuration changes must be made through Omni. If you want to access the Talos APIs you will need to re-provision the machine from Talos installation media.
```

The following section explains how to use a Machine Join Config to add your Talos nodes to Omni.

## Machine Join Config

> The Machine Join Config is a partial configuration that lets you register existing Talos nodes with Omni.

Reviewer: ```suggestion
The Machine Join Config is a configuration that lets you register existing Talos nodes with Omni.
```

Author: @Unix4ever's earlier review(one of the resolved comments) added the partial adjective @rothgar

> Use this approach for nodes you created independently, for example, with bare metal, Proxmox, or cloud VMs.

Reviewer: ```suggestion
Use this approach for nodes you booted machines from a standard Talos Linux installation media, for example, an ISO or VM disk.
```
A note that all of the providers will automatically provision connection information so they never need a separate join configuration.

<Tab title="Omni UI">
To download the Machine Join Config from the UI:

1. Log in to your **Omni dashboard**.
2. Click **Download Machine Join Config** to download the configuration file.

![Download Machine Join Config Button](./images/join-machine-machine-join-config.png)
</Tab>
<Tab title= "CLI">
To download the Machine Join Config using a join token:

1. Log in to your Omni dashboard.
2. Go to **Machine Management → Join Tokens**.

![Join Tokens Menu](./images/join-machines-machine-management.png)

3. Click the **⋯** button next to the token and select **Copy Token**.

![Copy Token](./images/join-machines-copy-token.png)

4. In your terminal, run the following command:

```bash
omnictl jointoken machine-config --token-id=<join-token>
```

Replace the `<join-token>` with the token you copied in step 3.
</Tab>
</Tabs>

### Supplying the Machine Join Config

The recommended way to use the join config is to provide it as userdata on your platform. For example:

- **Proxmox**: Use the NoCloud (cloud-init) datasource.

- **AWS / GCP / Azure**: Add it as instance user-data when launching a VM.

- **Bare metal / QEMU**: Provide it through NoCloud metadata or your provisioning tool.

> This ensures that the machine automatically joins Omni during its first boot.

Author: Andrey said I should not document this, as it would break when the machine is removed

> This ensures that the machine automatically joins Omni during its first boot.

Reviewer: Do we have any way to join a machine to Omni that doesn't have userdata?

---
title: Join machines to Omni
---

Omni lets you connect existing Talos nodes, whether they’re running on bare metal, in QEMU, or on cloud instances, to your Omni environment so it can automatically take over their management.

You can do this in two ways:

* **Machine Join Config**: Apply a configuration file to a Talos node to register it with Omni.

* **Kernel Parameters**: Add kernel parameters to your machine’s boot configuration so it automatically connects to Omni during startup.

> **Note:** Once a machine joins Omni, it is managed entirely through Omni.
> The local Talos API is disabled for security reasons, and all future configuration changes must be made via Omni.

The following sections explain each method in more detail and how to use them.

## Machine Join Config

> The Machine Join Config is a lightweight configuration file that lets you  register existing Talos nodes with Omni.

Reviewer 2: I guess I won't call it `lightweight configuration`. This is actually a partial config that should always stay with the machine. 

```suggestion
The Machine Join Config is a partial configuration that lets you register existing Talos nodes with Omni.
```

> Use this approach for nodes you created independently, for example, with `talosctl cluster create`, Proxmox, or cloud VMs.

Reviewer 2: I'd drop `talosctl cluster create` from here and add bare metal instead.

These nodes must be in **maintenance mode**, meaning no machine configuration (`controlplane.yaml` or `worker.yaml`) has been applied yet.

### How to use the Machine Join Config file

Follow these steps to download and apply the Machine Join Config file:

1. Log in to your Omni dashboard.

2. Select **Download Machine Join Config** to download the configuration file.

> ![Download Machine Join Config Button](./images/join-machine-machine-join-config.png)

Reviewer 2: You can also:
1. Download join configs for a specific token if you go to `Machine Management` -> `Join Tokens` -> `...` near the token.
2. We have CLI tools for downloading machine join configs, e.g.: `omnictl jointoken machine-config --token-id=6C8qmaEZeaaXcyF3qwerbyLFYWUqfWmM5cIc19c1slwA`

And the same story for kernel args.

> 3. Apply the configuration to your Talos nodes by running:

Reviewer 3: I would rather remove it, primary way should be to supply this join config as userdata on your specific platform: nocloud for Proxmox, userdata on AWS/GCP/..., etc.

This apply-config would break when the machine is removed.
